### ISSUE TITLE
BUG: Inversion operator (~) removes array in single element bool array

### ISSUE THREAD

**Issue by AlexanderKeijzer** [numpy/numpy#22812]
### Describe the issue:

The inversion operator (~) removes the numpy array in single element bool array and turns the result into a python boolean.

### Reproduce the code example:

```python
import numpy as np
criteria = np.array([False])
print("Result of ~[False]:")
print(~criteria)
print("Expectation:")
print(np.array([True]))
```

### NumPy/Python version information:

1.23.4 3.8.8 (default, Apr 13 2021, 19:58:26) 
[GCC 7.3.0]

### Context for the issue:

This causes issues when using this operation when indexing as follows:
```
import numpy as np
# Note, in a real usecase this element length is not consistently 1
data = np.array([2])
criteria = np.array([False])

selection = data[~criteria]
```
Instead of indexing with (array) [True] we now index with just True which adds an extra dimension, which is inconsistent compared to the same code where the element length is bigger than 1.

**Comment by shreyass-ranganatha** [comment]
This, seems to have been fixed in 1.23.5. (I'm not sure if this exists in 1.23.4, because the code block you've posted doesn't show the outputs and neither do I have the version installed to verify)

```python
>>> c = np.array([False])
>>> 
>>> ~c
array([ True])
>>> 
>>> ~c == np.array([True])
array([ True])
>>> 
>>> 
>>> a = np.array([2])
>>> a[~c]
array([2])
```

**Comment by seberg** [comment]
I first thought you mean returning a scalar rather than 0-D array, but @shreyass-ranganatha is right: all your example do the right thing, and I don't see that ever changed for many years.

How did you install your NumPy, @AlexanderKeijzer?!

**Comment by seberg** [comment]
@AlexanderKeijzer a bit of an early ping, but could you follow up on what exactly you are seeing?  I cannot reproduce the behavior with your examples, so right now I somewhat expect that there is something additional going on.
Unfortunately, there is no way to figure it out for us and I can't help but wonder if the issue is actually with non-NumPy objects.  (E.g. I am aware that torch has indexing issues which might run into something like this if mixed with NumPy.)

**Comment by AlexanderKeijzer** [comment]
@seberg I'm sorry, I've been a bit busy the last days. I've installed numpy using anaconda. The problem showed up in a project where I was converting from PyTorch to numpy arrays but I could reproduce it with just numpy. I'll give a more detailed reproduction tomorrow.

**Comment by AlexanderKeijzer** [comment]
@seberg So, I did have some time now at work to check this once more and it seems like I made a mistake here using squeeze instead of flatten (obfuscated a bit by transforming from torch to numpy in the middle).
In pure numpy I was doing this:
```python
import numpy as np
data = np.array([2])
criteria = np.array([[False]]).squeeze()
print(criteria) # array(False)
selection = data[~criteria]
print(selection) # array([[2]])
```
I must say, this indexing with a 0-dim array is a bit unintuitive and prone to these issues but probably known behavior. The inversion operator does implicitly convert a 0-dim bool array into a python primitive bool.

So yeah, happy to close this. Sorry!
